File: exam-conductor/archiveDCR/libs/exampen-proto/typescript/src/stroke.ts
/** Stroke domain types: raw chunks, processed strokes, upload status. */

import type { ChunkBindingStatus, UploadPath } from "./enums";

/** A single chunk of raw stroke data uploaded from the hub. */
export interface StrokeChunkUploadRequest {
  exam_id: string;
  pen_mac: string;
  chunk_index: number;
  total_chunks: number;
  payload_base64: string;
  checksum_crc32: string;
  upload_path: UploadPath;
  idempotency_key: string;
  binding_status?: ChunkBindingStatus;
}

/** Server acknowledgement for a received stroke chunk. */
export interface IngestAck {
  exam_id: string;
  pen_mac: string;
  chunk_index: number;
  accepted: boolean;
  deduplicated: boolean;
  next_expected_chunk: number;
  pen_upload_complete?: boolean;
}

/** Per-pen upload reconciliation state. */
export interface PenUploadStatus {
  pen_mac: string;
  acked_chunks: number[];
  total_chunks: number;
  complete: boolean;
}

/** Upload progress for all pens in an exam. */
export interface ExamUploadStatus {
  exam_id: string;
  pens: PenUploadStatus[];
}

/** Mapping of a stroke segment to a question on a page. */
export interface PageAssignment {
  page_number: number;
  question_id: string;
  point_count: number;
}

/** NATS event: raw stroke chunk ingested from hub. */
export interface StrokeRawEvent {
  event_id: string;
  event_type: "stroke.raw";
  event_version: "1.0.0";
  occurred_at: string;
  exam_id: string;
  pen_mac: string;
  chunk_index: number;
  total_chunks: number;
  payload_base64: string;
  checksum_crc32: string;
  upload_path: UploadPath;
  binding_status?: ChunkBindingStatus;
}

/** NATS event: stroke data normalized, deduplicated, committed. */
export interface StrokeProcessedEvent {
  event_id: string;
  event_type: "stroke.processed";
  event_version: "1.0.0";
  occurred_at: string;
  exam_id: string;
  pen_mac: string;
  student_id?: string;
  normalized_stroke_uri: string;
  page_assignments: PageAssignment[];
}

File: exam-conductor/archiveDCR/libs/exampen-proto/python/exampen_proto/stroke.py
"""Stroke domain models: raw chunks, processed strokes, upload status."""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel

from .enums import ChunkBindingStatus, UploadPath


class StrokeChunkUploadRequest(BaseModel):
    """A single chunk of raw stroke data uploaded from the hub."""

    exam_id: UUID
    pen_mac: str
    chunk_index: int
    total_chunks: int
    payload_base64: str
    checksum_crc32: str
    upload_path: UploadPath
    idempotency_key: str
    binding_status: Optional[ChunkBindingStatus] = None


class IngestAck(BaseModel):
    """Server acknowledgement for a received stroke chunk."""

    exam_id: UUID
    pen_mac: str
    chunk_index: int
    accepted: bool
    deduplicated: bool
    next_expected_chunk: int
    pen_upload_complete: Optional[bool] = None


class PenUploadStatus(BaseModel):
    """Per-pen upload reconciliation state."""

    pen_mac: str
    acked_chunks: list[int]
    total_chunks: int
    complete: bool


class ExamUploadStatus(BaseModel):
    """Upload progress for all pens in an exam."""

    exam_id: UUID
    pens: list[PenUploadStatus]


class PageAssignment(BaseModel):
    """Mapping of a stroke segment to a question on a page."""

    page_number: int
    question_id: str
    point_count: int


class StrokeRawEvent(BaseModel):
    """NATS event: raw stroke chunk ingested from hub."""

    event_id: str
    event_type: str = "stroke.raw"
    event_version: str = "1.0.0"
    occurred_at: datetime
    exam_id: UUID
    pen_mac: str
    chunk_index: int
    total_chunks: int
    payload_base64: str
    checksum_crc32: str
    upload_path: UploadPath
    binding_status: Optional[ChunkBindingStatus] = None


class StrokeProcessedEvent(BaseModel):
    """NATS event: stroke data normalized, deduplicated, and committed."""

    event_id: str
    event_type: str = "stroke.processed"
    event_version: str = "1.0.0"
    occurred_at: datetime
    exam_id: UUID
    pen_mac: str
    student_id: Optional[str] = None
    normalized_stroke_uri: str
    page_assignments: list[PageAssignment]
